design/acme-orders-challenges-crd.md

```diff
@@ -0,0 +1,388 @@
+// The possible options here map to the corresponding values in the
+// ACME specification.
+// Full details of these values can be found there.
+// Clients utilising this type **must** also gracefully handle unknown
+// values, as the contents of this enumeration may be added to over time.
+type State string
+
+const (
+	// Unknown is not a real state as part of the ACME spec.
+	// It is used to represent an unrecognised value.
+	Unknown State = ""
+
+	// Valid signifies that an ACME resource is in a valid state.
+	// If an Order is marked 'valid', all validations on that Order
+    // have been completed successfully and the order has been finalized
+    // with a CSR.
+    // You should perform a GET on the 'CertificateURL' in order to retrieve
+    // the existing certificate.
+	// This is a final state.
+	Valid State = "valid"
+
+	// Ready signifies that an ACME resource is in a ready state.
+    // If an Order is marked 'Ready', the order is ready to be finalized.
+    // The order must be finalized with a CSR, and once finalization succeeds,
+    // the order will transition into a 'valid' state.
+	// This is a transient state.
+	Ready State = "ready"
+
+	// Pending signifies that an ACME resource is still pending and is not yet ready.
+	// If an Order is marked 'Pending', the validations for that Order are still in progress.
+	// This is a transient state.
+	Pending State = "pending"
+
+	// Processing signifies that an ACME resource is being processed by the server.
+	// If an Order is marked 'Processing', the validations for that Order are currently being processed.
+	// This is a transient state.
+	Processing State = "processing"
+
+	// Failed signifies that an ACME resource has failed for some reason.
+	// If an Order is marked 'Failed', one of its validations may have failed for some reason.
+	// This is a final state.
+	Failed State = "failed"
+
+	// Expired signifies that an ACME resource has expired.
+	// If an Order is marked 'Expired', one of its validations may have expired or the Order itself.
+	// This is a final state.
+	Expired State = "expired"
+)
+```
+
+##### Challenge
+
+```go
+// +genclient
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +kubebuilder:resource:path=challenges
+type Challenge struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   ChallengeSpec   `json:"spec"`
+	Status ChallengeStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ChallengeList is a list of Challenges
+type ChallengeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []Challenge `json:"items"`
+}
+
+type ChallengeSpec struct {
+	AuthzURL string `json:"authzURL"`
+	Type     string `json:"type"`
+	URL      string `json:"url"`
+	DNSName  string `json:"dnsName"`
+	Token    string `json:"token"`
+	Key      string `json:"key"`
```

Comment:
What is the `Key` field of a challenge? The key authorization? Might deserve a comment if so. "key" is a super overloaded term in ACME/PKI :laughing: